### PR TITLE
Explain: Add arity and column type attributes.

### PR DESCRIPTION
**Pull request by wangandi**
Allows new explain plans to display arities and column types.

In order to allow the arity attribute to share code with `MirRelationExpr::arity`, there is a new helper method `arity_with_input_arities` that incrementally calculates arities. 
The method `typ_with_input_types` has been split into two methods: `col_with_input_cols` and `keys_with_input_keys` so that the column type attribute can share code with `MirRelationExpr::typ`. The future unique keys attribute will call `keys_with_input_keys`.

### Motivation

  * This PR adds a known-desirable feature.

Part of #14315. Ticks off the arity and column types.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### CODE REVIEW

File: src/expr/src/relation/mod.rs [hunk 1]
```diff
@@ -300,79 +300,211 @@ impl MirRelationExpr {
...
+                input_types.next().unwrap().clone()
+            }
+            Union { .. } => {
+                let mut result = input_types.next().unwrap().clone();
+                for input_col_types in input_types {
+                    for (base_col, col) in result.iter_mut().zip_eq(input_col_types) {
+                        *base_col = base_col
+                            .union(&col)
+                            .map_err(|e| format!("{}\nIn {:#?}", e, self))
+                            .unwrap();
```
Comment (aalexandrov) [inline src/expr/src/relation/mod.rs:410]: Urgh, didn't realize that we could panic here...

File: src/expr/src/relation/mod.rs [hunk 2]
```diff
@@ -300,79 +300,211 @@ impl MirRelationExpr {
...
+            }
+        }
+    }
+
+    /// Reports the unique keys of the relation given the arities and the unique
+    /// keys of the input relations.
+    ///
+    /// `input_arities` and `input_keys` are required to contain the
+    /// corresponding info for the input relations of
+    /// the current relation in the same order as they are visited by `try_visit1`
```
Comment (aalexandrov) [inline src/expr/src/relation/mod.rs:423]: nit: `try_visit1` does not exist any more, it's now called `try_visit_post`.

**Review by aalexandrov — APPROVED** [review]
Small nits, most controversially requesting to change the keyword for that attribute to `types` (which already exists in the code) and change the attribute name to `RelationType` (see comment for justification), otherwise feel free to merge.

File: src/repr/src/explain_new.rs [hunk 3]
```diff
@@ -359,6 +359,10 @@ impl From<RecursionLimitError> for ExplainError {
 /// A set of options for controlling the output of [`Explain`] implementations.
 #[derive(Debug)]
 pub struct ExplainConfig {
+    /// Show the number of columns.
+    pub arity: bool,
+    /// For each column, show the scalar type and whether the column is nullable
+    pub column_types: bool,
     /// Render implemented MIR `Join` nodes in a way which reflects the implementation.
     pub join_impls: bool,
     /// Show the `non_negative` in the explanation if it is supported by the backing IR.
```
[the remark this answers is not in the document]
Comment (wangandi) [reply src/repr/src/explain_new.rs:368]: I guess I did not rebase against main in a while.

[the remark this answers is not in the document]
Comment (wangandi) [reply src/repr/src/explain_new.rs:368]: I see it now that I've rebased.

File: src/repr/src/explain_new.rs [hunk 4]
```diff
@@ -359,6 +359,10 @@ impl From<RecursionLimitError> for ExplainError {
 /// A set of options for controlling the output of [`Explain`] implementations.
 #[derive(Debug)]
 pub struct ExplainConfig {
+    /// Show the number of columns.
+    pub arity: bool,
+    /// For each column, show the scalar type and whether the column is nullable
+    pub column_types: bool,
```
[the remark this answers is not in the document]
Comment (wangandi) [reply src/repr/src/explain_new.rs:365]: Oops, I did not noticed. Fixed!

**Pull request merged**